programs/mango-v4/src/state/orderbook/book.rs

```diff
@@ -128,6 +129,52 @@ impl<'a> Orderbook<'a> {
             remaining_quote_lots -= match_quote_lots;
             assert!(remaining_quote_lots >= 0);
 
+            let self_trade_behavior = match order.params {
+                OrderParams::Market {
+                    self_trade_behavior,
+                } => Some(self_trade_behavior),
+                OrderParams::ImmediateOrCancel {
+                    self_trade_behavior,
+                    ..
+                } => Some(self_trade_behavior),
+                _ => None,
+            };
+
+            let mut maker_fee = market.maker_fee;
+            let mut taker_fee = market.taker_fee;
+            let order_would_self_trade = mango_account_pk.eq(&best_opposing.node.owner);
+            if order_would_self_trade {
+                if let Some(self_trade_behavior) = self_trade_behavior {
+                    require!(
+                        self_trade_behavior != SelfTradeBehavior::AbortTransaction,
+                        MangoError::WouldSelfTrade
+                    );
+
+                    if self_trade_behavior == SelfTradeBehavior::CancelProvide {
+                        let event = OutEvent::new(
+                            other_side,
+                            best_opposing.node.owner_slot,
+                            now_ts,
+                            event_queue.header.seq_num,
+                            best_opposing.node.owner,
+                            best_opposing.node.quantity,
+                        );
+                        event_queue.push_back(cast(event)).unwrap();
+                        // TODO track cancelled order deletes differently
+                        matched_order_deletes
+                            .push((best_opposing.handle.order_tree, best_opposing.node.key));
+
+                        continue;
```

Comment:
At this point `remaining_base_lots` and `remaining_quote_lots` have already been changed. That should only happen after the self-trade check.